### PR TITLE
Clear scheduled timeout if channel is closed with incomplete WebSocket handshake

### PR DESCRIPTION
**Pull request by divijvaidya**
**Motivation:**

Consider a scenario when the client iniitiates a WebSocket handshake but before the handshake is complete,
the channel is closed due to some reason. In such scenario, the handshake timeout scheduled on the executor
is not cleared. The reason it is not cleared is because in such cases the handshakePromise is not completed.

**Modifications:**

This change completes the handshakePromise exceptinoally on channelInactive callback, if it has not been
completed so far. This triggers the callback on completion of the promise which clears the timeout scheduled
on the executor.

This PR also adds a test case which reproduces the scenario described above. The test case fails before the
fix is added and succeeds when the fix is applied.

**Result:**

After this change, the timeout scheduled on the executor will be cleared, thus freeing up thread resources.

**ICLA**
Signed the ICLA.

**Testing**
Tested using `mvn test` on `codec-http` module.

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientProtocolHandshakeHandler.java
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientProtocolHandshakeHandler.java:75]: Lets first check if it’s done yet and only if this is not true try to fail it. This will reduce the overhead in most cases as we will not create the extra exception

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by amizurov** [comment]
Hmm little bit confused, according current implementation if a `channel` will break during handshake the `ClientHandshakeStateEvent.HANDSHAKE_TIMEOUT` will be fired at the end of `handshakeTimeoutMillis` 
even if we don't cancel it and business logic can react on this event. Now if we silently cancel the `timeoutFeature` we never know about connection issue without extra logic since we do not have access from outside to `handshakePromise`.

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java
Comment (amizurov) [inline codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java:207]: Sorry maybe I miss something, how server can send close websocket frame if handshake is not complete yet.

**Comment by divijvaidya** [comment]
> Hmm little bit confused, according current implementation if a `channel` will break during handshake the `ClientHandshakeStateEvent.HANDSHAKE_TIMEOUT` will be fired at the end of `handshakeTimeoutMillis`
> even if we don't cancel it.

Yes, the callback to send `HANDSHAKE_TIMEOUT` event will be fired but the event will not be sent because the handler might have been removed from the pipeline at that time i.e. `ctx.isRemoved()` will be `true` in the code block where we sent the `HANDSHAKE_TIMEOUT`. In such cases, no-one will intercept the `HANDSHAKE_TIMEOUT` event. This code change will save the resources (namely executor thread) spent in tracking and executing the timeout callback which is useless after the handler has been removed.

Next we will discuss what can be the possible cases when the handler has been removed before the timeout callback is invoked. Let me provide an example for this.

I have a client implementation which uses SSL Handler with WebSocket. In my client, if SSL handshake fails for some reason, it leads to channel close which deactivates/deregisters the WebSocket handler. I have a logic at the application that is waiting for either `ClientHandshakeStateEvent.HANDSHAKE_TIMEOUT` or `ClientHandshakeStateEvent.HANDSHAKE_COMPLETE` but that logic is left waiting and neither events arrive. On closer inspection, I discovered that although the callback to send the `HANDSHAKE_TIMEOUT` was triggered but the event was never propagated since the handler was removed because channel was closed. This code change does not solve this scenario (I will file a separate PR for that) but I wanted to call it out as an example which is replicated by the unit test.

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientProtocolHandshakeHandler.java
Comment (divijvaidya) [reply codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientProtocolHandshakeHandler.java]: Done.

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java
Comment (divijvaidya) [reply codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java]: You are right. This is a bad example of a scenario for a channel close. I have changed the comment to remove this description. However, apart from this incorrect example, there can be number of cases when the channel might be closed before the handshake is complete. I have provide an example where SSL handshake fails below in the other comment.

**Comment by amizurov** [comment]
> > Hmm little bit confused, according current implementation if a `channel` will break during handshake the `ClientHandshakeStateEvent.HANDSHAKE_TIMEOUT` will be fired at the end of `handshakeTimeoutMillis`
> > even if we don't cancel it.
> 
> Yes, the callback to send `HANDSHAKE_TIMEOUT` event will be fired but the event will not be sent because the handler might have been removed from the pipeline at that time i.e. `ctx.isRemoved()` will be `true` in the code block where we sent the `HANDSHAKE_TIMEOUT`. In such cases, no-one will intercept the `HANDSHAKE_TIMEOUT` event. This code change will save the resources (namely executor thread) spent in tracking and executing the timeout callback which is useless after the handler has been removed.
> 
> Next we will discuss what can be the possible cases when the handler has been removed before the timeout callback is invoked. Let me provide an example for this.
> 
> I have a client implementation which uses SSL Handler with WebSocket. In my client, if SSL handshake fails for some reason, it leads to channel close which deactivates/deregisters the WebSocket handler. I have a logic at the application that is waiting for either `ClientHandshakeStateEvent.HANDSHAKE_TIMEOUT` or `ClientHandshakeStateEvent.HANDSHAKE_COMPLETE` but that logic is left waiting and neither events arrive. On closer inspection, I discovered that although the callback to send the `HANDSHAKE_TIMEOUT` was triggered but the event was never propagated since the handler was removed because channel was closed. This code change does not solve this scenario (I will file a separate PR for that) but I wanted to call it out as an example which is replicated by the unit test.

Yes and No. Event will be sent but handled only in `TailContext` where it will be freed.
You are right that we can save resources and not do this, and I am full of hands for this.
About your scenario you can add custom handler to pipeline and keep handshake state according to `ClientHandshakeStateEvent` even if channel will be closed you known in which state it happened.

**Comment by divijvaidya** [comment]
> You are right that we can save resources and not do this, and I am full of hands for this.

Does this mean you are ok with this change? If yes, would you approve it or shall I wait for Norman's approval?

> About your scenario you can add custom handler to pipeline and keep handshake state according to ClientHandshakeStateEvent even if channel will be closed you known in which state it happened.

That is how I have worked around this (i.e. introduce a custom handler with a promise which completes on handshake complete/timeout event or when the channel becomes inactive). But IMO, we could improve the default behaviour of Netty since it is not very evident from the default behaviour that one should not listen synchronously for handshake to complete without handling the case of channel close. I have proposed a few suggestions in https://github.com/netty/netty/issues/10515. I would appreciate if you could add your thoughts there. We can further continue the discussion about this scenario on that issue.

**Comment by divijvaidya** [comment]
@normanmaurer @amizurov Just checking if you need anything else from my side on this PR? I am currently waiting for review.

**Review by amizurov — APPROVED** [review]

**Comment by amizurov** [comment]
> @normanmaurer @amizurov Just checking if you need anything else from my side on this PR? I am currently waiting for review.

LGTM

**Comment by chrisvest** [comment]
@netty-bot test this please

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java:213]: Also call `clientChannel.finishAndReleaseAll()` and `serverChannel.finishAndReleaseAll()` otherwise we have leaks

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java
Comment (divijvaidya) [reply codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java]: updated

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Review by normanmaurer — APPROVED** [review]

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
@divijvaidya thanks a lot

**Pull request merged**
